packages/core/src/server.ts

```diff
@@ -0,0 +1,27 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: @loopback/core
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import { Context, Binding, Constructor } from '@loopback/context';
+
+export class Server extends Context {
```

Comment:
`Server` should be an interface.

When using a base class, a component implementing custom server/transport will need to depend on `@loopback/core` in order to get this base class. Because there may be multiple instances of `@loopback/core` in `node_modules` tree, we could end up with `Server` subclasses inheriting from different `Server` base class. This will break `instanceof` checks and probably TypeScript compile-time checks too. TypeScript interfaces are different, they are used at compile-time only and compared by content (member signatures), not by reference.

The difficulty I am foreseeing is how to say that `Server` should be also a `Context`, because we don't have interfaces defined for `Context`, `Binding` and other types required by `Context` API.  I guess it's time to create them?

Another problem: let's say `websocket` transport (a community maintained extension) depends on `@loopback/context@4.0.0` but core has moved to `@loopback/context@5.0.0` since then and the person maintaining the websocket transport did not update yet; and let's say that `@loopback/context@5.0.0` added new methods to `Context` or `Binding` classes. Because `WebsocketTransport` is using the older version of context under the hood, it will create a class instance that does not have these new methods. 

 1. In the better case, typescript will catch this problem at compile time and refuse the to compile an app using latest loopback core with this hypothetical websocket transport.

 2. In the worse case, we will discover missing members at runtime only.

In both cases the application developer is out of luck (especially if core depends on the newly added Context API) - they cannot upgrade their app to the latest core until the community developer of the extension updates the extension too.

I think this is a situation we really want to avoid.

A possible solution that comes to my mind is to inject `Context` from the core run by the application into the Server provided by the extension. This means a custom `Server` can no longer inherit from `Context` though.

```ts
class WebsocketServer implements Server {
  // type signature and @inject annotations omitted for brevity
  constructor(app, options) {
    this.context = app.createChildContext();
  }

  // ...
}

// in Context
class Context {
  // ...
  createChildContext(): Context {
    return new Context(this);
  }
}
```

@raymondfeng @ritch thoughts?